File: example/1d-linear-convection/weno3/cpp/01j/src/flux/basic/EngquistOsherFlux.cpp
#include "flux/basic/EngquistOsherFlux.h"
#include <cmath>
#include <stdexcept>

namespace cfd {

Real EngquistOsherFlux::compute_flux(Real uL, Real uR, Real wave_speed) const {
    Real cp = 0.5 * (wave_speed + std::abs(wave_speed));
    Real cm = 0.5 * (wave_speed - std::abs(wave_speed));
    return cp * uL + cm * uR;
}

void EngquistOsherFlux::compute_flux_array(const Vector& uL, const Vector& uR, 
                                         Vector& flux, Real wave_speed) const {
    if (uL.size() != uR.size() || uL.size() != flux.size()) {
        throw std::invalid_argument("Flux array sizes do not match");
    }
    
    Real cp = 0.5 * (wave_speed + std::abs(wave_speed));
    Real cm = 0.5 * (wave_speed - std::abs(wave_speed));
    
    for (size_t i = 0; i < flux.size(); ++i) {
        flux[i] = cp * uL[i] + cm * uR[i];
    }
}

} // namespace cfd

File: example/1d-linear-convection/weno3/fortran/cfd/04f/src/flux/engquist_osher_flux.f90
! ==================== engquist_osher_flux.f90 ====================
! Engquist-Osher flux computation for OneFLOW-CFD
!
! Purpose: Compute numerical flux using Engquist-Osher scheme
! Reference: B. Engquist and S. Osher, "One-sided difference
!            approximations for nonlinear conservation laws,"
!            Math. Comp., 1981
! Author: OneFLOW-CFD Team
! Date: Created
! ================================================================

module engquist_osher_flux_module
    use kinds, only: dp
    use config_module, only: CfdConfigType
    use mesh_module, only: MeshType
    implicit none
    
    private
    public :: engquist_osher_flux
    
contains
    
    ! Engquist-Osher flux implementation
    subroutine engquist_osher_flux(q_face_left, q_face_right, flux, config, mesh)
        real(dp), intent(in) :: q_face_left(:), q_face_right(:)
        real(dp), intent(out) :: flux(:)
        type(CfdConfigType), intent(in) :: config
        type(MeshType), intent(in) :: mesh
        
        integer :: i
        real(dp) :: c, cp, cm, u_L, u_R
        
        c = config%wave_speed
        
        do i = 1, mesh%nnodes
            cp = 0.5_dp * (c + abs(c))
            cm = 0.5_dp * (c - abs(c))
            u_L = q_face_left(i)
            u_R = q_face_right(i)
            flux(i) = cp * u_L + cm * u_R
        end do
    end subroutine engquist_osher_flux
    
end module engquist_osher_flux_module